### ISSUE TITLE
Multi-class classification with qiskit's VQC

### ISSUE THREAD

**Issue by sri1905** [Qiskit/qiskit-machine-learning#251]
I am following the tutorial given in qiskit's website [Neural Network Classifier and Regressor][1]. In the first part, classification, the third section refers to qiskit's VQC library. Everything works fine with the given X and y where there are only two classes. I modified the X and y slightly to include four classes instead of two using the following lines:

    num_inputs = 2
    num_samples = 100
    
    X = 2*np.random.rand(num_samples, num_inputs) - 1
    y = np.random.choice([0,1,2,3], 100)
    
    y_one_hot = np.zeros(( num_samples, 4 ))
    for i in range(len(y)):
        y_one_hot[i, y[i]]=1

The rest of the code is untouched. VQC with ZZFeatureMap, RealAmplitudes ansatz, cross_entropy loss function and COBYLA() optimizer. 
However, when I try to fit with this new data, the classifier is only running for 5 iterations and the weights are not being changed at all. The loss or objection function's value is always calculated as "nan". 

There is a [similar question][2] I had posted about weights not being optimized with VQC, but then I thought it was because of my data or VQC's configuration. After trying this example, I realised it is clearly to do something with multiple classes and not just the classifier's configuration.
 
Please shine light on how to do multi-class classification using the qiskit's VQC library.

### TLDR: VQC not optimizing weights if multi-class classification. Code: [Colab Notebook][3]  

  [1]: https://qiskit.org/documentation/machine-learning/tutorials/02_neural_network_classifier_and_regressor.html
  [2]: https://quantumcomputing.stackexchange.com/questions/21568/qiskits-classifier-is-not-optimising-the-weights
  [3]: https://colab.research.google.com/drive/1diZB4EUqaM8Ku4b2y7zbjKnxIxK_7hEv?usp=sharing

**Comment by jonvet** [comment]
I looked a bit into this. I you run your script there is a warning that gives a hint: 
```/Users/jonas/repos/qiskit-machine-learning/qiskit_machine_learning/utils/loss_functions/loss_functions.py:158: RuntimeWarning: divide by zero encountered in log2
  val = -np.einsum("ij,ij->i", target, np.log2(predict)).reshape(-1, 1)
```
looking at `predict` I saw that it's a (100, 4) array, but the 4th column is all zeros, leading to this warning and eventually the nans.

the zeros seem to be coming from here `qiskit_machine_learning.neural_networks.circuit_qnn.CircuitQNN._probabilities`. 
in particular from this loop:
```        
for i, circuit in enumerate(circuits):
    counts = result.get_counts(circuit)
    shots = sum(counts.values())

    # evaluate probabilities
    for b, v in counts.items():
        key = self._interpret(int(b, 2))
        if isinstance(key, Integral):
            key = (cast(int, key),)
        key = (i, *key)  # type: ignore
        prob[key] += v / shots
```
where key is either 0, 1 or 2 so the 3rd column of probs just stays 0. `key` is evaluated to 1 both when `b` is "01" and when it is "10".

**Comment by jonvet** [comment]
I can take a look at it if you assign me

**Comment by adekusar-drl** [comment]
@jonvet Sorry, I missed the comments. Feel free to take a look.

**Comment by sri1905** [comment]
Thank you for taking this up @jonvet. Meanwhile, are there any alternate ways to perform multiclass classification with the qiskit library?

**Comment by adekusar-drl** [comment]
@sri1905 Yes, I posted an example in the parallel issue you opened: https://github.com/Qiskit/qiskit-machine-learning/issues/256. Then in case of `CircuitQNN` you may pass a custom interpret function that maps measured outcome from the QNN to class labels. In case of OpflowQNN, custom observables need to be passed to `OpflowQNN`.

**Comment by sri1905** [comment]
>

**Comment by sri1905** [comment]
> @sri1905 Yes, I posted an example in the parallel issue you opened: #256. Then in case of `CircuitQNN` you may pass a custom interpret function that maps measured outcome from the QNN to class labels. In case of OpflowQNN, custom observables need to be passed to `OpflowQNN`.

Thank you. This helps a lot.

**Comment by sri1905** [comment]
Sorry, I closed it by mistake.

**Comment by adekusar-drl** [comment]
@jonvet Any updates on the issue? I can re-assign if you are busy.

**Comment by jonvet** [comment]
> @jonvet Any updates on the issue? I can re-assign if you are busy.

sorry for the delay, I will look into it this weekend

**Comment by sri1905** [comment]
@jonvet @adekusar-drl, I updated qiskit and qiskit_machine_learning using `pip install -U qiskit qiskit_machine_learning`. Even after this, I am still getting the same issue for multiple classes. The objective_function_vals (losses) are all **_nan_** and weights are **_constant_**. 
How can I get the version of qiskit and qiskit_machine_learning that you have fixed ? Thank you

**Comment by adekusar-drl** [comment]
In my case:
```python
from qiskit_machine_learning.version import get_version_info
print(get_version_info())
```
I get `0.4.0`. This is a developer version.

Same for qiskit:
```python
from qiskit.version import get_version_info
print(get_version_info())
0.20.0
```

**Comment by sri1905** [comment]
The versions I have are:
```
qiskit==0.19.1
qiskit-machine-learning==0.3.0
```

As you have mentioned I can not get qiskit_machine_learning==0.4.0. However, I can manually install qiskit==0.20.0 but this causes dependency conflicts between qiskit-machine-learning and qiskit-terra:
> qiskit-machine-learning 0.3.0 requires qiskit-terra>=0.19.1, but you have qiskit-terra 0.15.1 which is incompatible.


If I try to install qiskit_terra>=0.19.1, I get a conflict between qiskit and qiskit-terra
> qiskit 0.20.0 requires qiskit-terra==0.15.1, but you have qiskit-terra 0.19.2 which is incompatible.

Couple of other issues, I wanted to bring to your notice were:
The following points have been observed in qiskit's example code demonstrating [Neural Network Classifier and Regressor](https://qiskit.org/documentation/machine-learning/tutorials/02_neural_network_classifier_and_regressor.html). The only modifications to the example code are mentioned below, rest of the parameters and configurations are same as the example (Binary classification using VQC with ZZFeatureMap, RealAmplitudes ansatz and cross_entropy loss function) 
1. If you use ADAM() optimiser instead of the COBYLA() given in the example, `vqc.fit(X, y_one_hot)` runs infinitely and the `callback_graph()` function doesn't even print the graph. Hence giving me no scope to debug further.
2. In the `callback_graph()` if you also store the weights in a list along with `obj_func_eval`, it can be observed that the `objective_func_vals` are reducing but the weights remain constant. 


Could you please let me know which versions I can use to perform multi-class classification without any errors ?

**Comment by adekusar-drl** [comment]
To make use of the fix, you have to install qiskit machine learning 0.4.0, a developer version, since this fix has not been released yet. This means you have to install it from the sources. So far, QML requires qiskit-terra >= 0.19.1 which was released some time ago. To get of the errors, I suggest to have a clean python environment without any qiskit packages. In this environment, please install qiskit via `pip install qiskit` first. Then, clone this repo via `git clone` and then in the project folder run `pip install .`. There's a dot in the command. This command installs QML from the cloned repo, from the recent sources, and you get QML 0.4.0 as I have.

Regarding the warnings:

> qiskit-machine-learning 0.3.0 requires qiskit-terra>=0.19.1, but you have qiskit-terra 0.15.1 which is incompatible.

Yes, this should be fixed if you wish to use QML 0.3.0. QML algorithms may not yield correct results with qiskit-terra 0.15.1. You can upgrade qiskit-terra by running `pip install -U qiskit`.

> qiskit 0.20.0 requires qiskit-terra==0.15.1, but you have qiskit-terra 0.19.2 which is incompatible.

In general, it is not a problem. Qiskit is a meta-package that contains qiskit-terra and qiskit-aer. One you have qiskit installed, you may upgrade qiskit-terra manually and get such warnings.

The rest of the problems depend on your setup and exact versions of the packages you have installed.

**Comment by sri1905** [comment]
@adekusar-drl @jonvet Thank you so much for all the clarification and help. Just one last thing, I am trying to train a large dataset with qiskit's classifiers. Could you please guide me to a resource which illustrates fundamentals of real world application of qiskit machine learning? I have not been able to find answers despite going the documentation and qiskit's youtube videos. The qiskit documentation and tutorials don't shine light on practical application and the qiskit youtube videos explain the background and mathematics but I am still left with some doubts regarding implementation.

**Comment by adekusar-drl** [comment]
I'm not sure I understood what you are looking for. What is "real world application"? A real dataset? Training on a real and especially large dataset may be pretty painful. It may be take a lot of time.

Take a look here: https://qiskit.org/textbook-beta/course/machine-learning-course/ This is a brand new chapter on quantum machine learning in the Qiskit textbook. May be you find something new.

**Comment by sri1905** [comment]
Sorry for generalising the terms. I meant a dataset with 1,000-10,000 datapoints and 100-1000 features. I have gone through the interactive course but yes, will take a look at it because I believe it recently got updated. Yes, it is very painful, using an aer simulator it took over 1hr to train 280 x 10 dataset on 10 qubits and the resulting accuracy was 22%

I am mainly looking for guidelines on optimising the way I am using qiskit for classification. Basic things like:
- Do number of qubits have to be same as features or is there a better way to encode more features on a feature map with fewer qubits ?
- Difference between simulators and is there a way to set the batch size, iterations and epochs ?

**Comment by adekusar-drl** [comment]
Oh, these are large numbers! Many samples may be tackled via `warm_start` parameter. You split a dataset into batches and train a model on each batch with `warm_start=True` sequentially, except for the first batch, of course. But this feature may be broken, a couple of issues are filed recently. 

Number of features can be reduced by applying PCA before training a model. But such a reduction may easily reduce performance of the model. You can try to apply amplitude encoding, e.g. use `RawFeatureVector` to load N classical features into log_2 (N) qubits. This encoding has its own problem, resulting circuit is very deep if the number of features is large.
In general, choice of a feature map is very important and there are many ways how to choose it. There's no silver bullet here.

Regarding the simulators, it is better to ask in another place. I'm not a good person to answer such questions.

**Comment by sri1905** [comment]
> Oh, these are large numbers! Many samples may be tackled via `warm_start` parameter. You split a dataset into batches and train a model on each batch with `warm_start=True` sequentially, except for the first batch, of course. But this feature may be broken, a couple of issues are filed recently.

Okay, I shall try it anyway.

> Number of features can be reduced by applying PCA before training a model. But such a reduction may easily reduce performance of the model. You can try to apply amplitude encoding, e.g. use `RawFeatureVector` to load N classical features into log_2 (N) qubits. This encoding has its own problem, resulting circuit is very deep if the number of features is large. In general, choice of a feature map is very important and there are many ways how to choose it. There's no silver bullet here.

Yes, I have been experimenting with various dimentionality reduction techniques. I will look into `RawFeatureVector`, thank you.

> Regarding the simulators, it is better to ask in another place. I'm not a good person to answer such questions.

Thank you very much !
